src/main/java/org/opensearch/ad/cluster/HashRing.java

```diff
@@ -197,16 +212,15 @@ public void buildCirclesOnAdVersionsDirectly() {
      *    and still send RCF request to it. If new node added during cooldown period, realtime job won't
      *    choose it as model partition owning node, thus we may have skewed load on data nodes.
      *
+     * [Important!]: When you call this function, make sure you TRY ACQUIRE adVersionCircleInProgress first.
+     *               Check {@link HashRing#buildCirclesOnAdVersions(ActionListener)} and
+     *               {@link HashRing#buildCirclesOnAdVersions(DiscoveryNodes.Delta, ActionListener)}
+     *
      * @param removedNodeIds removed node ids
      * @param addedNodeIds added node ids
      * @param actionListener action listener
      */
-    public void buildCirclesOnAdVersions(Set<String> removedNodeIds, Set<String> addedNodeIds, ActionListener<Boolean> actionListener) {
-        if (!adVersionCircleInProgress.tryAcquire()) {
-            LOG.info("AD version hash ring change in progress, return.");
-            actionListener.onResponse(false);
-            return;
-        }
+    private void buildCirclesOnAdVersions(Set<String> removedNodeIds, Set<String> addedNodeIds, ActionListener<Boolean> actionListener) {
```

Comment:
@kaituo , per your ask, I moved the `adVersionCircleInProgress.tryAcquire()` block to functions which calling buildCirclesOnAdVersions:  `HashRing#buildCirclesOnAdVersions(ActionListener)` and `HashRing#buildCirclesOnAdVersions(DiscoveryNodes.Delta, ActionListener)`

**Pros**: This can return earlier if doesn't get semaphore, so we can save some cost, like the set comparison.
**Cons**: If someone wants to build new function on top of buildCirclesOnAdVersions, should pay attention to acquire `adVersionCircleInProgress` semaphore first.

Reply:
Thanks for the accommodation.  The cons is troublesome indeed.  And current code build on demand on some code path, not for all of them.  This will increase maintenance cost and not efficient.  I wrote a draft version (forgive my grammar or naming error/discrepancy, just to show you what I have in mind).  Could you check? Specifically,

First, I removed buildCirclesOnAdVersions(DiscoveryNodes.Delta delta, ActionListener<Boolean> listener) and depend on buildCirclesOnAdVersions(ActionListener<Boolean> actionListener).  So we have one public build method that can be called in all of the places.  That's what the original HashRIng code does. This ease maintenance and avoids possible errors mentioned in your cons.
Second, the only difference between real time and historical is the cooldown.  Other things are the same.  So I check nodeChangeEvents.isEmpty() in buildCirclesOnAdVersions.
Third, I put the logic of cooldown logic inside buildAndGetOwningNodeRealTime. We still have 2 hash rings.  buildAndGetOwningNodeRealTime is in charge of updating real time hash ring. 
Fourth, do extra inProgress.availablePermits() >= 1 check inside buildCirclesOnAdVersions(Set<String> removedNodeIds, Set<String> addedNodeIds, ActionListener<Boolean> actionListener).  The objective is to fail fast and double check misuse of inProgress.

```
public class HashRing {
    ...
    private TreeMap<Version, TreeMap<Integer, DiscoveryNode>> circle, realTimeCircle;
    private Semaphore inProgress, realTimeInProgress;
    // the UTC epoch milliseconds of the most recent successful update
    private long lastUpdate, realTimeLastUpdate;
    ...
    private ConcurrentLinkedQueue<Boolean> nodeChangeEvents;

    public boolean buildCirclesOnAdVersions(ActionListener<Boolean> actionListener) {
        
        if (nodeChangeEvents.isEmpty() && !circle.isEmpty()) {
            return false;
        }

        if (!inProgress.tryAcquire()) {
            LOG.info("Hash ring change in progress, return.");
            return false;
        }

        DiscoveryNode[] eligibleDataNodes = nodeFilter.getEligibleDataNodes();
        Set<String> eligibleNodeIds = new HashSet<>();
        for (DiscoveryNode node : eligibleDataNodes) {
            eligibleNodeIds.add(node.getId());
        }
        Set<String> currentNodeIds = nodeAdVersions.keySet();
        Set<String> removedNodeIds = Sets.difference(currentNodeIds, eligibleNodeIds);
        Set<String> addedNodeIds = Sets.difference(eligibleNodeIds, currentNodeIds);
        buildCirclesOnAdVersions(removedNodeIds, addedNodeIds, actionListener);
    }

    private void buildCirclesOnAdVersions(Set<String> removedNodeIds, Set<String> addedNodeIds, ActionListener<Boolean> actionListener) {
       if (inProgress.availablePermits() >= 1) {
              // we need inProgress to have 0 permits.  If not, fail fast, throw exception
              ...
        }
        try {
            ...
                        int size = nodeChangeEvents.size();
                        TreeMap<Integer, DiscoveryNode> circle = null;
                        for (PluginInfo pluginInfo : plugins.getPluginInfos()) {
                            if (AD_PLUGIN_NAME.equals(pluginInfo.getName()) || AD_PLUGIN_NAME_FOR_TEST.equals(pluginInfo.getName())) {
                                Version version = ADVersionUtil.fromString(pluginInfo.getVersion());
                                circle = adVersionCircles.computeIfAbsent(version, key -> new TreeMap<>());
                                nodeAdVersions.put(curNode.getId(), ADVersionUtil.fromString(pluginInfo.getVersion()));
                                break;
                            }
                        }
                        if (circle != null) {
                            for (int i = 0; i < VIRTUAL_NODE_COUNT; i++) {
                                circle.put(Murmur3HashFunction.hash(curNode.getId() + i), curNode);
                            }
                        }
			            circle = newCircle;
			            while (size-- > 0) {
			                Boolean poll = nodeChangeEvents.poll();
			                if (poll == null) {
			                    break;
			                }
			            }
			            lastUpdate = clock.millis();
                    }
                }
                LOG.info("All nodes in AD version hash ring: {}", nodeAdVersions);

                ...
            }, e -> {
                adVersionCircleInProgress.release();
                actionListener.onFailure(e);
                LOG.error("Fail to get node info to build AD version hash ring", e);
            }));
        } catch (Exception e) {
            LOG.error("Failed to build AD version circles", e);
            adVersionCircleInProgress.release();
            actionListener.onFailure(e);
        }
    }

    private void buildCirclesOnAdVersionsDirectly() {
    	buildCirclesOnAdVersions(ActionListener.wrap(r->.., e->...))
    }

    /**
     * Compute the owning node of modelID using consistent hashing
     * @param key example: http-latency-rcf-1
     * @return the owning node of the modeID
     */
    public Optional<DiscoveryNode> buildAndGetOwningNodeRealTime(String key) {
        buildCirclesOnAdVersionsDirectly();

        if (realTimeInProgress.tryAcquire() && realTimeLastUpdate != lastUpdate && clock.millis() - realTimeLastUpdate > coolDownPeriod.getMillis()) {
            try {
                // recursive deep copy of circle
	        	realTimeCircle = ... 
	        	realTimeLastUpdate = lastUpdate
            } finally {
            	realTimeInProgress.release();
            }
        }

        return getOwningNode(realTimeCircle, key);
    }

    public Optional<DiscoveryNode> buildAndGetOwningNode(String key) {
        buildCirclesOnAdVersionsDirectly();

        return getOwningNode(circle, key);
    }

    public Optional<DiscoveryNode> getOwningNode(TreeMap<Integer, DiscoveryNode> hashRing, String key) {
    	int modelHash = Murmur3HashFunction.hash(key);
        Map.Entry<Integer, DiscoveryNode> entry = hashRing.higherEntry(modelHash);

        // The method can return an empty Optional. Say two concurrent getOwningNode requests to
        // the hash ring before it's been built. The first one starts building it,
        // turning on inProgress. The second one returns from build and continues on to
        // the rest of hashing and look up while the ring is still being built and thus empty.
        // The second getOwningNode request returns an empty Optional in this case.
        return Optional.ofNullable(Optional.ofNullable(entry).orElse(hashRing.firstEntry())).map(x -> x.getValue());
    }
}

```

Reply:
Summary of our offline discussion. Tuned hash ring method/variable name to make it clear and add semaphore check before building hash ring.
1. If we use two semaphore, the realtime may copy in middle of historical hash ring updating. That means some node may have less virtual node than other node in realtime hash ring. Then the load will be skewed for realtime as we depends on hash ring totally to distribute load.
2. We should update realtime hash ring on demand based on event queue. But for some edge cases like the node is busy with GC, the node change event may not be consumed, so the event queue is still empty, that means we may miss that node in hash ring for this case. For historical analysis, if we miss some node, it will run slower. So we check removed/added node every time to make sure the historical analysis is always up to date. And historical analysis will sleep for some time (5 seconds by default) between pieces, comparing with this sleep time, the performance loss of checking removed/added node can be ignored.